hathor/p2p/peer_connections.py

```diff
@@ -0,0 +1,200 @@
+        after getting connected and before getting READY.
+        """
+        assert addr not in self._connecting_outbound
+        assert addr in self._handshaking
+        assert addr not in self._ready
+        self._handshaking.pop(addr)
+
+    def on_ready(self, *, addr: PeerAddress, peer_id: PeerId) -> HathorProtocol | None:
+        """
+        Callback for when a connection gets to the READY state.
+        If the PeerId of this connection is duplicate, return the protocol that we should disconnect.
+        Return None otherwise.
+        """
+        assert addr not in self._connecting_outbound
+        assert addr in self._handshaking
+        assert addr not in self._ready
+
+        protocol = self._handshaking.pop(addr)
+        self._ready[addr] = protocol  # We always index it by address, even if its PeerId is duplicate.
+
+        connection_to_drop: HathorProtocol | None = None
+
+        # If there's an existing connection with the same PeerId, this is a duplicate connection
+        if old_connection := self.get_ready_peer_by_id(protocol.peer.id):
+            # We choose to drop either the new or the old connection.
+            if self._should_drop_new_connection(protocol):
+                # We return early when we drop the new connection,
+                # so we don't override the old connection in _addr_by_id with it below.
+                return protocol
+
+            # When dropping the old connection, we do override it in _addr_by_id below.
+            connection_to_drop = old_connection
+
+        self._addr_by_id[peer_id] = addr
+        return connection_to_drop
+
+    def on_ready_disconnect(self, *, addr: PeerAddress, peer_id: PeerId) -> None:
+        """Callback for when a connection is closed during the READY state."""
+        assert addr not in self._connecting_outbound
+        assert addr not in self._handshaking
+        assert addr in self._ready
+        self._ready.pop(addr)
+
+        if self._addr_by_id[peer_id] == addr:
+            self._addr_by_id.pop(peer_id)
+
+    def on_unknown_disconnect(self, *, addr: PeerAddress) -> None:
+        """Callback for when a connection is closed during an unknown state."""
+        assert addr not in self._handshaking
+        assert addr not in self._ready
+        if addr in self._connecting_outbound:
+            self._connecting_outbound.remove(addr)
+
+    @staticmethod
+    def _should_drop_new_connection(new_conn: HathorProtocol) -> bool:
+        """
+        When there are connections with duplicate PeerIds, determine which one should be dropped, the old or the new.
+        Return True if we should drop the new connection, and False otherwise.
+
+        The logic to determine this is `(my_peer_id > other_peer_id) XNOR new_conn.inbound`.
+        """
+        my_peer_is_larger = bytes(new_conn.my_peer.id) > bytes(new_conn.peer.id)
+        return my_peer_is_larger == new_conn.inbound
```

Comment:
I refactored this from `ConnectionsManager`, simplifying its logic to a simple boolean operation. The old docstring and comments were incorrect, because it stated a decision to drop either the connection initiated by us, or the one initiated by the other peer.

This was incorrect because it's possible that both duplicate connections are initiated by the other peer, for example. Therefore the correct description is just choosing whether to drop the old connection or the new, independent of who initiated them. The `new_conn.inbound` is used in the boolean logic but doesn't actually convey any meaning.